### ISSUE TITLE
Add Money.Subscription

### ISSUE THREAD

**Issue by maennchen** [kipcole9/money#58]
Continuing discussing from #55

**Comment by maennchen** [comment]
@kipcole9 

I think that there are two models on how to upgrade subscriptions.

1. deduct remaining money from the next subscription
2. add remaining money in from of time to next subscription

How we should calculate the second one without losing precision is currently not clear to me.

**Comment by kipcole9** [comment]
Yes, its an interesting discussion.  The Stripe model is basically:

> If the new subscription (downgrade or upgrade) has the same periodicity (same billing frequency and plan period) then the new plan comes into affect at the next billing date.  If the subscription has a new periodicity (moving from monthly to yearly for example) then a new billing period is started on the date the plan is effective with a pro-rated credit/debit applied to the new plan for the first period (or for as many periods as required to bring the payment stream into balance.

This seems a reasonable model.  But I can also understand organisations that would always want to start a new billing period on all plan changes.

How does that fit (or not) with your requirements?

**Comment by maennchen** [comment]
Since we have special requirements for payment (We have to implement five different payment providers), the only common denominator for recurring billing is that we can bill exactly the same amount every time.

Therefore we can not calculate the worth of your current subscription and deduct it from the first period of your new subscription.

This leaves us only with the option to give you extra time on the first period of the new subscription.

I have done this as follows until now (in pseudo code):

old_daily_price = total amount old sub / total days in period

remainder = old_daily_price * used days in period

new_daily_price = total amount new sub / total days in period

days_to_add = ceil(new_daily_price / remainder)

As I said before, I did not care about precision issues because I always rounded up the days.

**Comment by kipcole9** [comment]
Thanks for the clarify, I now understand your requirement - you need to keep the payment constant and therefore you are adding or subtracting billing units (days in your case) to reconcile.

Would you mind reviewing the following and letting me know if I've captured your needs correctly as part of the scope I'm planning to implement?

### A Plan

A plan defines the offer or product that a customer subscribes to.  The following attributes apply:

* Plan name
* Plan ID
* Interval: `day`, `week`, `month`, `year`
* Interval count: positive integer
* Subscription price: `Money.t`

### A Subscription

A subscription links a plan to a subscriber.  For the purposes of `Money` it contains:

* A plan
* The last billing date
* Billing currency (defaults to the plan currency)

### Plan Upgrade/Downgrade 

When a subscriber upgrades or downgrades a plan the following strategies can be applied:

* **Finish the current billing period before applying the new plan**.  The means that there is no need to deal with partially completed billing period and no prorating is required.  This is the default when the old and new plans have the same `interval` and `interval multiple`

* **Start the new plan immediately**.  In this case we have a potential partial credit of the old plan period.  There are two options that can applied in the case:

    * *Apply credit to the new plans first period* which will reduce the amount billed in the first period by the amount of credit left in the old plan.  This is an example of prorating the first periods billing amount.

    * Extend the subscription.  The first billing period is extended by an amount of value that the credit from the old plan represents in the new plan.  Then the billing period is reset to the normal period.  This is an example of prorating the first billing periods length.

**Comment by maennchen** [comment]
@kipcole9 This look almost exactly like what we built. 👍

**Comment by kipcole9** [comment]
Got it.  I'll work up a prototype over the weekend and then seek your feedback and suggestions again then.

**Comment by kipcole9** [comment]
Initial code has been [committed in the subscription branch](https://github.com/kipcole9/money/tree/subscription).  It compiles but I have not tried any execution yet so clearly its not ready.  I will add tests over the next 24 hours and correct identified bugs.  However I would welcome any code review, API suggestions and any other observations.

I said I would work up the code over the weekend and since its mid-morning Monday now where I am I wanted to meet my objective 😄 

The primary api is `Money.Subscription.change/3`.  See `h Money.Subscription` and `h Money.Subscription.change` in `iex`.

**Comment by kipcole9** [comment]
Some examples: (not verified the results are correct). Hopefully helps with any feedback or suggestions.
```elixir
iex> p1 =  %{interval: :day, interval_count: 30, price: #Money<:USD, 1>}
iex> p2 = %{interval: :day, interval_count: 30, price: #Money<:USD, 2>}

# Change plan at the end of the current billing period
iex> Money.Subscription.change p1, p2, ~D[2018-03-01]                           
%{
  carry_forward: #Money<:USD, 0>,
  credit_amount_applied: #Money<:USD, 0>,
  credit_period_applied: 0,
  following_billing_date: ~D[2018-04-30],
  next_billing_amount: #Money<:USD, 2>,
  next_billing_date: ~D[2018-03-31]
}

# Change plan part way through the plan period and credit the amount
iex> Money.Subscription.change p1, p2, ~D[2018-03-01], effective: ~D[2018-03-05]                  
%{
  carry_forward: #Money<:USD, 0>,
  credit_amount_applied: #Money<:USD, 0.87>,
  credit_period_applied: 0,
  following_billing_date: ~D[2018-04-04],
  next_billing_amount: #Money<:USD, 1.13>,
  next_billing_date: ~D[2018-03-05]
}

# Change plan part way through the plan period and extend the period
iex> Money.Subscription.change p1, p2, ~D[2018-03-01], effective: ~D[2018-03-05], prorate: :period
%{
  carry_forward: #Money<:USD, 0>,
  credit_amount_applied: #Money<:USD, 0.87>,
  credit_period_applied: 13,
  following_billing_date: ~D[2018-04-17],
  next_billing_amount: #Money<:USD, 2>,
  next_billing_date: ~D[2018-03-05]
}
```

**Comment by maennchen** [comment]
@kipcole9 Very cool! I’m currently checking the code by eye and will later compare with my existing implementation.

> This module calculates all subscription changes on the basis
  that billing is done in advance.

What is the reason for that paragraph?

**Comment by maennchen** [comment]
https://github.com/kipcole9/money/blob/b6c91a040b2a2dea0eb7d8af975a5165315052b4/lib/money/subscription.ex#L127

What happens if the new_plan does not contain any price? I would pattern match for the price in the function header.

**Comment by maennchen** [comment]
We should also check the currencies of the old and new price. All strategies but waiting for the end of the period do not allow different currencies.

**Comment by maennchen** [comment]
https://github.com/kipcole9/money/blob/b6c91a040b2a2dea0eb7d8af975a5165315052b4/lib/money/subscription.ex#L204

What happens if the old plan and new plan do have other interval periods?

Shouldn’t we always calculate with days? This would be way more precise.

**Comment by kipcole9** [comment]
Good comments and suggestions, thanks.  Definitely more work to be done!

1. The reason for the note on billing in advance is that when the credit is calculated is should be a credit against the account amount consumed.  If you bill in advance then the credit is the amount remaining in the current period.  If you bill in arrears then the credit is really a debit - since you haven't paid yet for the part you used. Not difficult to incorporate "in arrears" billing, just that I don't take that into account currently.

2. Definitely I will allow for the plan currencies to be different

3. Shouldn't all plans have a price? I will pattern match in the function header, makes sense and simplifies the code in several places.

**Comment by kipcole9** [comment]
Yes, you are right - I realised after I pushed the commit that working on fractions isn't the right way.  Next commit I'll fix that.

**Comment by maennchen** [comment]
https://github.com/kipcole9/money/blob/b6c91a040b2a2dea0eb7d8af975a5165315052b4/lib/money/subscription.ex#L252

Should we do that many date calculations ourselves? For example Timex.shift does all of that.

**Comment by maennchen** [comment]
> Definitely I will allow for the plan currencies to be different

So how will you calculate how much price / days to prorate without converting the price to the same currency?

Currently the code doesn’t care if I have 10 dollars, euros or apples. It only matters that it is ten.
If I get 10 yens prorated to a 10 dollar plan, this is a huge difference.

**Comment by kipcole9** [comment]
I would definitely have to convert to the same base currency otherwise as you say, no comparison is valid.  On reflect though, I think that perhaps normalising a plan to a common base currency may be better left as user requirement.  I should check that the currencies are the same and error if not.  Would agree?  Or is converting to a base currency here a reasonable thing to do?

**Comment by kipcole9** [comment]
On date calculations:  I'd prefer not to add another dependency.  In additional, Timex was written before Elixir 1.5 which introduces the mechanics to do the calculation without having to add a library.  The calculations on dates are quite simply (times are much more complex but not applicable - at least for now).  Lastly, I don't believe Timex supports non-ISO calendars.  This code supports any calendar that supports the Calendar behaviour.

**Comment by maennchen** [comment]
@kipcole9 I would let the user do that. This will have implications on the books (you need to save the exchange rate for example). So I would error if there are different currencies.

**Comment by kipcole9** [comment]
Yes, I am aware of that;  I [have a PR submitted for Elixir](https://github.com/elixir-lang/elixir/pull/7417) to add `months_in_year/1` to allow for that.  So [you can see](https://github.com/kipcole9/money/blob/subscription/lib/money/subscription.ex#L281-L287) that I try to call that function on the calendar module if it exists (which it won't for now) and then assume 12 if not. 

The same issue applies to the [Balinese saka calendar](https://en.wikipedia.org/wiki/Balinese_saka_calendar) also has an intercalary month from time-to-time.

**Comment by maennchen** [comment]
@kipcole9 Ah, this was you? I thought that it was interesting that I saw something covering this just today.

Now let’s assume that my plan is 48 months long in a hebrew calendar. There‘s a good chance for a leap year. So we have 3 times 12 months and 1 time 13 months. I don‘t see how you handle that condition. As I see it, you only ask once for the months in a year based on the start date.

**Comment by kipcole9** [comment]
Good catch - I will fix the logic.  I have added the pattern match to `change/4` to ensure there is a price, and that the currencies match.  One down, several to go. I'm on a long flight tonight so probably expect an more material update in about 12 hours.

**Comment by kipcole9** [comment]
BTW, that issue is only applicable if the plan is in "months" in the current logic.  If its "days" or "years" it should calculate correctly.  Months, meaning "3rd of each month" obviously requires different logic - including the logic you noted.

**Comment by kipcole9** [comment]
Updated in the subscription branch

1. All credit amounts and periods now based upon the "price per day" calculation
2. Correctly calculate monthly plans over a year in length according to the calendar being applied (not fixed to 12 months in a year)

Next step is to work on tests.

**Comment by maennchen** [comment]
Very cool. Based on this i will compare results with the code we‘ve written. Maybe I also have a few tests to copy.

**Comment by maennchen** [comment]
I just tested against our ported tests.

I have two findings:

1. We need the `start_date` of the subscription. I mean the date after the prorate and before the lifetime is applied. I called it `billing_period_start_date` and put it commented out in our tests.

2. I have a different result when changing a subscription on the day it was bought. I'll have to calculate again because I'm not sure if the problem was in our code or in yours.

Those are the tests that I've ported:

```elixir
defmodule MoneyTest do
  @moduledoc false

  use ExUnit.Case

  describe "change/4" do
    test "should get days to add to subscription upgrade" do
      today = ~D[2018-02-14]

      old = %{
        price: Money.new(:CHF, Decimal.new(130)),
        interval: :month,
        interval_count: 3,
      }
      new = %{
        price: Money.new(:CHF, Decimal.new(180)),
        interval: :month,
        interval_count: 6,
      }

      assert %{
        carry_forward: Money.zero(:CHF),
        credit_amount_applied: Money.new(:CHF, Decimal.new("67.20")),
        credit_days_applied: 68,
        # billing_period_start_date: ~D[2018-04-21],
        following_billing_date: ~D[2018-10-21],
        next_billing_amount: new.price,
        next_billing_date: today
        } == Money.Subscription.change(old, new, ~D[2018-01-01], effective: today, prorate: :period)
    end

    test "should get days to add to subscription upgrade when upgrading the same day" do
      today = ~D[2018-01-01]

      old = %{
        price: Money.new(:CHF, Decimal.new(130)),
        interval: :month,
        interval_count: 3,
      }
      new = %{
        price: Money.new(:CHF, Decimal.new(180)),
        interval: :month,
        interval_count: 6,
      }

      assert %{
        carry_forward: Money.zero(:CHF),
        credit_amount_applied: Money.new(:CHF, Decimal.new("130.00")),
        credit_days_applied: 131,
        # billing_period_start_date: ~D[2018-05-11],
        following_billing_date: ~D[2018-11-11],
        next_billing_amount: new.price,
        next_billing_date: today
        } == Money.Subscription.change(old, new, ~D[2018-01-01], effective: today, prorate: :period)
    end

    test "should get at least 1 day when subscription upgrade is from a very low subscription to a very high one" do
      today = ~D[2018-01-14]

      old = %{
        price: Money.new(:CHF, Decimal.new("0.5")),
        interval: :month,
        interval_count: 1,
      }
      new = %{
        price: Money.new(:CHF, Decimal.new(1000)),
        interval: :month,
        interval_count: 36,
      }

      assert %{
        carry_forward: Money.zero(:CHF),
        credit_amount_applied: Money.new(:CHF, Decimal.new("0.30")),
        credit_days_applied: 1,
        # billing_period_start_date: ~D[2018-01-15],
        following_billing_date: ~D[2021-01-15],
        next_billing_amount: new.price,
        next_billing_date: today
        } == Money.Subscription.change(old, new, ~D[2018-01-01], effective: today, prorate: :period)
    end
  end
end
```

**Comment by kipcole9** [comment]
Really appreciate the collaboration.  I need a few hours sleep before I can get into the code again but very likely I still have bugs.  I need to write a lot more tests yet and your contribution really helps a lot.  Thank you.

**Comment by kipcole9** [comment]
I'd like to clarify my understanding:
```elixir
billing_period_start_date = last_billing_date + credit_days_applied
```
Is that correct?  I also wonder if you would accept `credit_period_ends` instead of `billing_period_start_date`?  It feels to me that it better captures the event.  Of course `credit_period_ends` is one day earlier than `billing_period_start_date`.

**I note that we have a one day difference between your calculation and mine**.  This is due to my implementation simply adding the `credit_days_applied`giving, in the second test above, as `~D[2018-01-12]` for me.  My assumption is that the `effective date` has been consumed and therefore the credit starts tomorrow.  Your assumption is, I think, that today also counts as a credit day.  What is your view of the best way to handle this?  I can see it both ways and I'm not sure what the expected practise is.

I'm now working out why, for the second test, I get a different date by 2 days from your calculation of `following_billing_date`.  1 day of the difference is because of the comment above, the other I'm working on now.

**Comment by maennchen** [comment]
> Is that correct?

Yes. I thin the idea with credit_period_ends is good ant that wirks for me.

> My assumption is that the effective date has been consumed and therefore the credit starts tomorrow. 

Hmm... I never really thought about this. I can‘t give you any arguments in favor or against this practice.

> I'm now working out why, for the second test, I get a different date by 2 days from your calculation of following_billing_date. 

Keep in mind that this problem could very well be a problem of our implementation and not of this library.

**Comment by kipcole9** [comment]
The second of your tests returns:
```elixir
  1) test change/4 should get days to add to subscription upgrade when upgrading the same day (MoneySubscriptionTest)
     test/subscription_test.exs:57
     Assertion with == failed
     code:  assert %{carry_forward: Money.zero(:CHF), credit_amount_applied: Money.new(:CHF, Decimal.new("130.00")), credit_days_applied: 131, credit_period_ends: ~D"2018-05-11", following_billing_date: ~D"2018-11-11", next_billing_amount: new.price(), next_billing_date: today} == Money.Subscription.change(old, new, today, effective: today, prorate: :period)
     left:  %{carry_forward: #Money<:CHF, 0>, credit_amount_applied: #Money<:CHF, 130.00>, credit_days_applied: 131, credit_period_ends: ~D[2018-05-11], next_billing_amount: #Money<:CHF, 180>, next_billing_date: ~D[2018-01-01], following_billing_date: ~D[2018-11-11]}
     right: %{carry_forward: #Money<:CHF, 0>, credit_amount_applied: #Money<:CHF, 130.00>, credit_days_applied: 131, credit_period_ends: ~D[2018-05-11], next_billing_amount: #Money<:CHF, 180>, next_billing_date: ~D[2018-01-01], following_billing_date: ~D[2018-11-09]}
     stacktrace:
       test/subscription_test.exs:71: (test)
```
As I look at it, the calculation is:
```elixir
# next billing date is correctly calculated according to
# your understanding and mine
next_billing_date = ~D[2018-01-01]

# Add six months being the period of the plan
following_billing_date = ~D[2018-07-01]

# Now add 131 days of the credit being applied
Money.Subscription.add_days following_billing_date, 131
~D[2018-11-09]
```
I note your test expects `~D[2018-11-11]` which I can't currently figure out how to derive.

**Comment by kipcole9** [comment]
I've pushed the latest code to the GitHub repo and included your tests - please let me know if thats ok.  If not I will of course remove the tests.

**Comment by maennchen** [comment]
Maybe I should tackle this problem since I‘m the one that has access to the other code and can compare directly.

**Comment by maennchen** [comment]
Of course you may use my tests. 👍

**Comment by kipcole9** [comment]
I have pushed another few commits that improve documentation and updates the changelog.  I also reviewed the failing tests and I think my calculations are correct so I adjusted the tests which now pass.  **I'm not certain yet** and I still need to do some substantial work on additional tests.

**Comment by kipcole9** [comment]
My confidence in the calculations is increasing as I add more tests but I need to test some more edge cases.

The current plan is to update the README to include a section on subscriptions and add some additional tests leading to release on hex this weekend.

Of course I'm still after feedback in order to maximise the fit of this module to your requirements.  Especially any cases where `change/4` doesn't delivery the correct result.

BTW my pull request for `Calendar.months_in_year/1` has been accepted and will ship in Elixir 1.7.  I do testing on 1.5, 1.6 and 1.7-dev before release.

**Comment by maennchen** [comment]
@kipcole9 Thanks a lot for your work!

I‘ll be checking this one maybe-wrong result on friday. I don‘t really have the time to do it today.

I would integrate the new module into our existing code also at friday. This way I can tell you if everything worked as expected.

**Comment by kipcole9** [comment]
Much appreciated. Of course no rush - real work comes first.

**Comment by maennchen** [comment]
```elixir
iex> Money.Subscription.next_billing_date(%{interval: :month, interval_count: 1}, ~D[2018-01-31])
** (MatchError) no match of right hand side value: {:error, :invalid_date}
    (ex_money) lib/money/subscription.ex:442: Money.Subscription.next_billing_date/2
```

**Comment by maennchen** [comment]
I also think we have another problem:

If I book a monthly subscription on the 31 of january and we renew two times, the billing date will be `2018-02-28` and the next `2018-03-28`. It should be `2018-03-31` though.

We always provided the first billing date to our `calculate_new_period_dates` to correct for that.

**Comment by kipcole9** [comment]
Got it, thank you.  Fixed the [first part](https://github.com/kipcole9/money/commit/2970b096bd30e3cc4cd85cd5e584784ed8fbb01d) but not yet the second.

**Comment by kipcole9** [comment]
And now the second part is, I think fixed.

I have done quite a lot of refactoring (apologies for the elapsed time to do this)

1. A `Subscription` now exists as a thing.  `Money.Subscription.change_plan/3` when called with a `Subscription.t` will maintain a list of plans and the change record that is applied to it

2. Subscription can be created with `Subscription.new/1` but any map with the relevant fields can be used.

I changed a lot of the field names to better reflect that this is not about billing per-se, but more about when interval starts.

I think this is now a stable API.  I still need to add more tests related to changing plans with a subscription which I'll do over the next few days.

If you have the time your feedback would be most welcome.

**Comment by maennchen** [comment]
@kipcole9 Great! I did some testing and found a small issue with the options. I created the PR #59 for that.

Other than that, everything works perfect for me.

**Comment by kipcole9** [comment]
Subject to your feedback I think the module is ready for release.  I made a couple of small API changes to be consistent with the rest of Money.

* Most functions now return `{:ok, data}` or `{:error, {exception, message}}`
* For those functions there is a `!` version that returns the data or raises
* I added additional tests for Subscriptions
* Additions to README and CHANGELOG

I'm soon on a flight for a few hours and will do another code review.  Your comments are most welcome.  This has taken far longer than I anticipated but I'm happy with the subscription and plan abstractions so hopefully they hold value for you and others.
